kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/source/SourceContextReader.scala

```diff
@@ -39,7 +40,10 @@ object SourceContextReader {
       offsetMap <- Try(context().offsetStorageReader.offset(key).asScala).toOption.filterNot(_ == null)
       path      <- offsetMap.get("path").collect { case value: String => value }
       line      <- offsetMap.get("line").collect { case value: String if value forall Character.isDigit => value.toInt }
-    } yield sourceRoot.withPath(path).atLine(line)
+      inst <- offsetMap.get("inst").collect {
+        case value: String if value forall Character.isDigit => Instant.ofEpochMilli(value.toLong)
+      }
+    } yield sourceRoot.withPath(path).atLine(line, inst)
```

Comment:
The connector "watermark" is file path + line, and the timestamp (last modified/ created). `.atLine` taking the timestamp implies that the line is linked to a timestamp. I am not saying it is not a way to keep the state, but it means each line could have a different timestamp. If it was like:

`withPath().atLine(line).withTimestamp(inst)` I think it will avoid the potential implication I mention above.

Reply:
And this is the comment which sent me on a 2 day detour and refactored the world.  But it  has simplified so much.